presto-main/src/main/java/com/facebook/presto/catalogserver/CatalogServer.java

```diff
@@ -24,69 +24,117 @@
+        this.getTableHandleCache = cacheBuilder.build(CacheLoader.from(this::loadGetTableHandle));
+        this.listTablesCache = cacheBuilder.build(CacheLoader.from(this::loadListTables));
+        this.listViewsCache = cacheBuilder.build(CacheLoader.from(this::loadListViews));
+        this.getViewsCache = cacheBuilder.build(CacheLoader.from(this::loadGetViews));
+        this.getViewCache = cacheBuilder.build(CacheLoader.from(this::loadGetView));
+        this.getMaterializedViewCache = cacheBuilder.build(CacheLoader.from(this::loadGetMaterializedView));
+        this.getReferencedMaterializedViewsCache = cacheBuilder.build(CacheLoader.from(this::loadGetReferencedMaterializedViews));
     }
 
+    /*
+        Metadata Manager Methods
+     */
+
     @ThriftMethod
     public boolean schemaExists(TransactionInfo transactionInfo, SessionRepresentation session, CatalogSchemaName schema)
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        return metadataProvider.schemaExists(session.toSession(sessionPropertyManager), schema);
+        return get(schemaExistsCache, new CacheKey<>(session, schema));
     }
 
     @ThriftMethod
     public boolean catalogExists(TransactionInfo transactionInfo, SessionRepresentation session, String catalogName)
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        return metadataProvider.catalogExists(session.toSession(sessionPropertyManager), catalogName);
+        return get(catalogExistsCache, new CacheKey(session, catalogName));
     }
 
     @ThriftMethod
     public String listSchemaNames(TransactionInfo transactionInfo, SessionRepresentation session, String catalogName)
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        List<String> schemaNames = metadataProvider.listSchemaNames(session.toSession(sessionPropertyManager), catalogName);
-        if (!schemaNames.isEmpty()) {
-            return writeValueAsString(schemaNames, objectMapper);
-        }
-        return EMPTY_STRING;
+        List<String> schemaNames = get(listSchemaNamesCache, new CacheKey(session, catalogName));
+        return schemaNames.isEmpty()
+                ? EMPTY_STRING
+                : writeValueAsString(schemaNames, objectMapper);
     }
 
     @ThriftMethod
     public String getTableHandle(TransactionInfo transactionInfo, SessionRepresentation session, QualifiedObjectName table)
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        Optional<TableHandle> tableHandle = metadataProvider.getTableHandle(session.toSession(sessionPropertyManager), table);
+        CacheKey cacheKey = new CacheKey(session, table);
+        Optional<TableHandle> tableHandle = get(getTableHandleCache, cacheKey);
+        if (!tableHandle.isPresent()) {
+            getTableHandleCache.refresh(cacheKey);
+            tableHandle = get(getTableHandleCache, cacheKey);
+        }
```

Comment:
We don't have to do this. What if
1. we refreshed
2.  there are suddenly 10K request coming and drain the cache
3. and when we hit `get`, the cache is already gone

This is a typical multi threading racing condition.

If we check the java doc of loading cache's `get`, it states: "Returns the value associated with key in this cache, first loading that value if necessary". So, do we really need to refresh?

Reply:
So the reason that I have this double check is due to two reasons
1. There is many cases where the original cache call during catalog creation will miss with an empty result but very shortly after would be populated. This meant that trying to access several tables would give us an empty result when the table actually exists
2. Since that would hinder the testing of the caching because, currently the cache has simple caching without data consistency so theres no way for me to update that current cached value without doing this.

Possibly a TODO could be added to remove this check once we add updated cached data checking?

Reply:
I'm not sure I understand the logic here
1. why is the process not taken care of by `get` method?
2. why do we have this additional logic only for `getTableHandleCache` but not the rest of cache

Reply:
1. It is taken care of in the get method. So, initially I thought we would be able to test the caching at the very least with the fact that it doesn't consider data consistency as of now. But it seems like during the HiveQueryRunner startup there are cached metadata that is already outdated when creating the catalogs before any queries are made. This check is the least that I could modify to make it testable, thus the double check for now (of course if would be removed once we consider if the cached result is accurate or not.
2. It is only done on this because its the only one that had this problem ^

I can remove the check however, it would affect the ability to test the cache as of now at the very least